### ISSUE TITLE
Try add covariance for containers

### ISSUE THREAD

**Issue by ZetaNumbers** [fitzgen/bumpalo#170]
Enables a small(?) but valid set of cases. May come with a small penalty, until we have `Unique` built-in. Tho it seems like even in std `Unique<T>` does not enable appropriate optimizations currently, based on my experiments with `std::vec::Vec`. On the other hand `Box` enables "ownership optimizations", maybe due to `#[lang = "..."]` attribute.

Make sure to benchmark difference.

**Comment by fitzgen** [comment]
I doubt that lifetime variance has any effect on optimizations. AFAIK that's all erased by the time code goes to LLVM, and I'd expect that any `rustc` optimizations for `Box` are looking at the lang attributes.

**Comment by ZetaNumbers** [comment]
> I doubt that lifetime variance has any effect on optimizations.

Of course, it's just that there's currently no nonaliased covariant pointer for rust, so the only option is maybe aliased covariant `NonNull<T>`, vs. nonaliased but invariant `&mut ManuallyDrop<T>`

**Comment by douglas-raillard-arm** [comment]
I just bumped into this issue while replacing a `std::boxed::Box` inside a recursive enum like this one:
```
struct Value<'a> {
    Str(&'a str),
    Wrapped(Box<Value<'a>>)
} 
```

The consuming code rightfully expects `Value<'a>` to be covariant in `'a`, but that breaks with bumpalo currently.
